overlord/confdbstate/confdbmgr.go

```diff
@@ -114,52 +114,109 @@ func (m *ConfdbManager) clearOngoingTransaction(t *state.Task, _ *tomb.Tomb) err
+	WriteTxID string   `json:"write-tx-id,omitempty"`
+}
+
+// addReadTransaction adds a read transaction for the specified confdb, if no
+// write transactions is ongoing.
+func addReadTransaction(st *state.State, account, confdbName, id string) error {
+	txs, updateFunc, err := getOngoingTxs(st, account, confdbName)
 	if err != nil {
-		if !errors.Is(err, &state.NoStateError{}) {
-			return err
-		}
+		return err
+	}
 
-		commitTasks = make(map[string]string, 1)
+	if txs == nil {
+		txs = &confdbTransactions{}
 	}
 
-	confdbRef := account + "/" + confdbName
-	if taskID, ok := commitTasks[confdbRef]; ok {
-		return fmt.Errorf("internal error: cannot set task %q as ongoing commit task for confdb %s: already have %q", commitTaskID, confdbRef, taskID)
+	if txs.WriteTxID != "" {
+		return fmt.Errorf("cannot read confdb (%s/%s): a write transaction is ongoing", account, confdbName)
 	}
 
-	commitTasks[confdbRef] = commitTaskID
-	st.Set("confdb-commit-tasks", commitTasks)
+	txs.ReadTxIDs = append(txs.ReadTxIDs, id)
+	updateFunc(txs)
 	return nil
 }
 
-func unsetOngoingTransaction(st *state.State, account, confdbName string) error {
-	var commitTasks map[string]string
-	err := st.Get("confdb-commit-tasks", &commitTasks)
+// setWriteTransaction sets a write transaction for the specified confdb schema,
+// if no other transactions (read or write) are ongoing.
+func setWriteTransaction(st *state.State, account, schemaName, id string) error {
+	txs, updateFunc, err := getOngoingTxs(st, account, schemaName)
 	if err != nil {
-		if errors.Is(err, &state.NoStateError{}) {
-			// already unset, nothing to do
-			return nil
+		return err
+	}
+
+	if txs == nil {
+		txs = &confdbTransactions{}
+	}
+
+	if txs.WriteTxID != "" || len(txs.ReadTxIDs) != 0 {
+		op := "read"
+		if txs.WriteTxID != "" {
+			op = "write"
 		}
+
+		return fmt.Errorf("cannot write confdb (%s/%s): a %s transaction is ongoing", account, schemaName, op)
+	}
+
+	txs.WriteTxID = id
+	updateFunc(txs)
+	return nil
+}
+
+func getOngoingTxs(st *state.State, account, schemaName string) (*confdbTransactions, func(*confdbTransactions), error) {
+	var confdbTxs map[string]*confdbTransactions
+	err := st.Get("confdb-ongoing-txs", &confdbTxs)
+	if err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, nil, err
+		}
+
+		confdbTxs = make(map[string]*confdbTransactions, 1)
+	}
+
+	ref := account + "/" + schemaName
+	updateFunc := func(ongoingTxs *confdbTransactions) {
+		if ongoingTxs == nil || (ongoingTxs.WriteTxID == "" && len(ongoingTxs.ReadTxIDs) == 0) {
+			delete(confdbTxs, ref)
+		} else {
+			confdbTxs[ref] = ongoingTxs
+		}
+
+		if len(confdbTxs) == 0 {
+			st.Set("confdb-ongoing-txs", nil)
+		} else {
+			st.Set("confdb-ongoing-txs", confdbTxs)
+		}
+	}
+	return confdbTxs[ref], updateFunc, nil
+}
+
+func unsetOngoingTransaction(st *state.State, account, schemaName, id string) error {
+	txs, updateFunc, err := getOngoingTxs(st, account, schemaName)
```

Comment:
```suggestion
	txs, updateTransactionState, err := getOngoingTxs(st, account, schemaName)
```

Maybe a name more like that? Unsure, but `updateFunc` isn't super clear without digging into what the returned closure does.

Reply:
I like it, thanks!